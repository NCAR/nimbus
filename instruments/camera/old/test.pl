#!/usr/bin/perl
$i = 0;
my @array = (1,2,3,4,5,6);
splice @array, 2, 1;
splice @array, $i, 1;
print join(":", @array);
print "\n";
