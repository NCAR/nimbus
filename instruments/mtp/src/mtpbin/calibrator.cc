#include "calibrator.h"
#include <iostream>

using namespace std;

Calibrator::Calibrator(std::vector<float> cnd0, std::vector<float> gof, std::vector<float> gec1, std::vector<float> gec2) 
{
  _CND0 = cnd0;
  _GOF = gof;
  _GEC1 = gec1;
  _GEC2 = gec2;

  // Create a gain vector with all elements set to zero.
  for (int i=0; i<NUM_CHANNELS; i++) _Gain.push_back(0.0);

  // Set constants for platinum wire gain equation for temperature of target. 
  // These should remain constant as long as physical target doesn't change.
  // If target is replaced, these may change.
  _AA = -244.3364635;
  _Bb = 0.462418;
  _cC = 0.0000588;
  _DD = -0.000000013;
  _Wtg = 0.1;

  // the angle associated with the horizontal scan
  _LocHor = 5; 

  return;
}

std::vector<float> Calibrator::Calibrate(std::vector<int> ScanCounts, std::vector<int> PltWireCnts, std::vector<int> TargetCnts, float OAT)
{
  // The equations in this routine come from MTPGainEquation.docx (in svn)
  std::vector<float> ScanBTs;

  std::vector<float> Gnd;
  // Create a Gnd vector with all elements set to zero.
  for (int i=0; i<NUM_CHANNELS; i++) Gnd.push_back(0.0);

  int TargIndx = 0;   	// Index into ScanCounts for the target w/ND on
  int TargNDIndx = 3;  // Index into ScanCounts for the target w/ND off
  for (int i=0; i<NUM_CHANNELS; i++)
  {
    Gnd[i] = (TargetCnts[TargIndx]-TargetCnts[TargNDIndx])/_CND0[i];
    TargIndx = TargIndx+1;
    TargNDIndx = TargNDIndx+1;
  }

  // Counts for TMix
  float r = 350.0+(250.0*(PltWireCnts[4]-PltWireCnts[0])/(PltWireCnts[7]-PltWireCnts[0])); 
  
  // Temperature of the mixer (TMix, also known as Tifa)
  float TMix = _AA +_Bb*r + _cC*r*r + _DD*r*r*r;

  // g() is initialized to a value at the beginning of a flight and then is adjusted on
  // a scan by scan basis in an iterative fashion.  
  for (int i=0; i<NUM_CHANNELS; i++)
  {
    if (_Gain[i] == 0.0) {
      // Initialize the gain
      _Gain[i] = (TMix-_GOF[0])*_GEC2[i] + _GEC1[i];
    } else {
      _Gain[i] = (_Gain[i] + Gnd[i] * _Wtg)/ (1+_Wtg);
    }
    cout << "Gain["<<i<<"]="<<_Gain[i]<<"\n";
  }

  cout << "Calculate Tb...\n";
  // Calculate Brightness Temperatures
  for (int i=0; i<NUM_CHANNELS;i++)
  {
    for (int j=0;j<NUM_SCAN_ANGLES;j++)
    {
        ScanBTs.push_back(OAT + (ScanCounts[i*10+j] - ScanCounts[i*10+_LocHor]) / _Gain[i]);
    }
  }

  return ScanBTs;
}
