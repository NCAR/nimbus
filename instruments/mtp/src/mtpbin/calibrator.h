/*! \file calibrator.h
 *  \brief This class encapsulates the calibration of the MTP scans.  
 ** 
 **
 ** The Microwave Temperature Profiler (MTP) instrument build by JPL for
 ** NCAR as part of the set of instruments known as HAIS (build specifically
 ** in support of the Haiper GV aircraft) came with processing software 
 ** written in Visual Basic 6.0. 
 **
 ** Copyright 2015-2016 University Corporation for Atmospheric Research
 **  - VB6 and Algorithm Copyright MJ Mahoney, NASA Jet Propulsion Laboratory
 **
 */
#ifndef _CALIB_h
#define _CALIB_h

#include "mtp.h"
#include <vector>
#include <string> 


/*! \class Calibrator
 *
 * This class encapsulates the calibration of the MTP scans.  It makes use of 
 * constants that are written into the .REF file for the flight in question.
 * I am unclear as to when that file gets written.
 *
 * When the MTP instrument completes a scan of the atmosphere the scan counts
 * are converted to Brightness Temperatures using this class.
 * 
 */
class Calibrator
{

public:

  /**
   * Constructor
   * @param cnd0 std::vector<float> is the 'Cnd0' values from the REF file.  
   * I believe that these are the "temperatures" of the target in each channel
   * in degrees C when the noise diode is on.  Comments in the VB code call
   * it the Noise Diode Temperature Fit Offset value.
   * @param gof std::vector<float> is the 'GOF' values from the REF file.
   * I believe that these are an offset for the temperature of the mixer, one
   * for each channel.  Comments in the code say they are Gain Equation Offsets
   * but their use belies that.
   * @param gec1 std::vector<float> is the 'GEC(chan,1)' vector from the REF file
   * called Gain Equation Coefficients, they do seem to be used in that way
   * There is one value per channel, this one is the offset coefficient.
   * @param gec2 std::vector<float> is the 'GEC(chan,2)' vector from the REF file
   * called Gain Equation Coefficients, this one is the linear coefficient
   * applied to the difference of the Mixer temperature minus the GOF above.
   */
  Calibrator(std::vector<float>, std::vector<float>, std::vector<float>, std::vector<float>);

  /**
   * Calibrate method
   * @param ScanCounts std::vector<int> is the vector of counts produced by the MTP
   * instrument for this scan. This vector should be of length 36 - three
   * points for each angle (one per channel) and six points for the target -
   * three per channel for the noise diode turned on and for it turned off.
   * @param PltWireCnts is the platinum wire counts.
   * @param TargetCnts is the counts from the target. First 3 are with ND on, next 3
   * off.
   * @param OAT is outside air temperature from the A line of the RAW MTP file (Kelvin).
   *
   * @returns a std::vector<float> of Brightness Temperatures for the scan. 
   * There are 30 elements to the vector, 3 values at each of the 10 angles
   * of the MTP's scan.
   */
  std::vector<float> Calibrate(std::vector<int>, std::vector<int>, std::vector<int>,float);
  

private: 

  ///< For storing the rolling gain value - one value per channel
  std::vector<float> _Gain;

  ///< Constants for platinum wire gain equation for temperature of target
  float _AA;
  float _Bb;
  float _cC;
  float _DD;
  float _Wtg;

  ///< The scan angle associated with the horizontal scan.
  int _LocHor;

  ///< Elements that are pulled from the flight's REF file.  
  ///< e.g. "c:\MTP\Data\NGV\HIPPO-5\20110822\NG20110822.REF"
  std::vector<float> _CND0;
  std::vector<float> _GOF;
  std::vector<float> _GEC1;
  std::vector<float> _GEC2;
  
};

#endif
