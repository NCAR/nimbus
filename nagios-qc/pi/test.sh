#! / bin / bash           
output=$(psg radar)
echo "$output"
