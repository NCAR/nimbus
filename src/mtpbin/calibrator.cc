#include "calibrator.h"
#include <iostream>

using namespace std;

Calibrator::Calibrator(std::vector<float> cnd0, std::vector<float> gof, std::vector<float> gec1, std::vector<float> gec2) 
{
  _CND0 = cnd0;
  _GOF = gof;
  _GEC1 = gec1;
  _GEC2 = gec2;

  // Create a gain vector with all elements set to zero.
  for (int i=0; i<NUM_CHANNELS; i++) _Gain.push_back(0.0);

  // Set constants for platinum wire gain equation for temperature of target.
  _AA = -244.3364635;
  _Bb = 0.462418;
  _cC = 0.0000588;
  _DD = -0.000000013;
  _Wtg = 0.1;

  return;
}

std::vector<float> Calibrator::Calibrate(std::vector<int> ScanCounts, std::vector<int> PltWireCnts, std::vector<int> TargetCnts)
{
  std::vector<float> ScanBTs;

  std::vector<float> Gnd;
  // Create a Gnd vector with all elements set to zero.
  for (int i=0; i<NUM_CHANNELS; i++) Gnd.push_back(0.0);

  int TargIndx = 0;   	// Index into ScanCounts for the target w/ND on
  int TargNDIndx = 3;  // Index into ScanCounts for the target w/ND off
  for (int i=0; i<NUM_CHANNELS; i++)
  {
    Gnd[i] = (TargetCnts[TargIndx]-TargetCnts[TargNDIndx])/_CND0[i];
    TargIndx = TargIndx+1;
    TargNDIndx = TargNDIndx+1;
  }

  // Counts for TMix
  // float TMC = PltWireCnts[4];
  float r = 350.0+(250.0*(PltWireCnts[4]-PltWireCnts[0])/(PltWireCnts[7]-PltWireCnts[0])); 
  
  // Temperature of the mixer (TMix, also known as Tifa)
  //float TMix = _AA +_Bb*TMC + _cC*TMC*TMC + _DD*TMC*TMC*TMC;
  float TMix = _AA +_Bb*r + _cC*r*r + _DD*r*r*r;

  // g() is initialized to a value at the beginning of a flight and then is adjusted on
  // a scan by scan basis in an iterative fashion.  
  for (int i=0; i<NUM_CHANNELS; i++)
  {
    if (_Gain[i] == 0.0) {
      // Initialize the gain
      //_Gain[i] = TMix*_GEC2[i];
      _Gain[i] = (TMix-_GOF[0])*_GEC2[i] + _GEC1[i];
    } else {
      _Gain[i] = (_Gain[i] + Gnd[i] * _Wtg)/ (1+_Wtg);
    }
    cout << "Gain["<<i<<"]="<<_Gain[i]<<"\n";
  }

  cout << "Calculate Tb...\n";
  // Calculate Brightness Temperatures
  float OAT = 222.53; // Kelvin
  int LocHor = 5; // the angle associated with the horizontal scan
  for (int j=0;j<NUM_SCAN_ANGLES;j++)
  {
    for (int i=0; i<NUM_CHANNELS;i++)
    {
        ScanBTs.push_back(OAT + (ScanCounts[i+j*3] - ScanCounts[i+LocHor*3]) / _Gain[i]); // Target 1
    }
  }
  for (int i=0; i<NUM_CHANNELS;i++)
  {
    for (int j=0;j<NUM_SCAN_ANGLES;j++)
    {
	cout << "Tb="<<OAT + (ScanCounts[i+j*3] - ScanCounts[i+LocHor*3])/_Gain[i]<<"\n";
    }
  }
  cout << "Done\n\n";

  return ScanBTs;
}
