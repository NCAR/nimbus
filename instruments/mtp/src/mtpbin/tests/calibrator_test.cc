#include <iostream>
#include <fstream>
#include <stdlib.h>

#include "../calibrator.h"
#include "test.h"

using namespace std;

/*
 * This program is designed to test the Calibrator class to assure that 
 * it is performing as expected.  The current output file:
 * calibrator.out has been vetted to assure that it has correct information
 * resulting from the tests this program peforms.
 *
 * Usage:
 * calibrator_test > calibrator_test.new
 * diff calibrator_test.out calibrator_test.new
 *
 * There should be no differences found.  If differences are found then 
 * something has affected the functionality of the class and should be 
 * investigated and resolved.
 */

main () {

  // Put together a sample case for testing the calibrator 
  // Get constants from test.h

  // Initialize a calibrator instance.
  Calibrator CBrator(cnd0, gof, gec1, gec2);

  // Call the calibrator function to convert counts to Tb.
  // Calculate Tb from counts
  std::vector<float> ScanBTs = CBrator.Calibrate(ScanCounts, PltWireCnts,TargetCnts, OAT);

  // Write generated Tb to screen.
  cout<<"ScanBTs:";
  for (int i=0; i<NUM_BRT_TEMPS;i++)
  {
      cout<< ScanBTs[i] << ",";
  }
  cout<<"\n";
}
