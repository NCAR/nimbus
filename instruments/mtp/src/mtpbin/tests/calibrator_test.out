Gain[0]=19.9552
Gain[1]=23.1852
Gain[2]=26.1352
Calculate Tb...
ScanBTs:214.913,216.166,217.068,218.671,220.526,222.53,224.033,225.537,227.942,229.896,216.405,216.88,217.699,219.338,220.891,222.53,223.048,224.169,225.204,226.412,219.086,219.584,219.928,220.77,221.688,222.53,223.525,224.328,225.438,226.433,
